[Joeie Tao] iP

joeietao opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [(here)](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

joeietao committed:
completed level 1: greet, echo, exit

joeietao committed:
added Level-2: Add, List

joeietao committed:
added Task.java class

joeietao committed:
added markAsDone, markAsUndone, returnDescription to Task.java

joeietao committed:
updated Task.java, including markAsDone, markAsUndone and returnDescription

joeietao committed:
added Level 3: Mark as done to Duke.java (users can now mark their tasks as done/undone)

joeietao committed:
added Deadline.java, Event.java and Todo.java as children classes of Task.java

joeietao committed:
added level-4: ToDos, Events, Deadlines (users can now only add 3 types of tasks, Todos, Events and Deadlines)

joeietao committed:
removed returnDescription from Task.java

joeietao committed:
corrected the print statement for list

joeietao committed:
A-TextUiTesting: Automated Text UI Testing

joeietao committed:
Level 5: exceptions (added Duke Exception)

joeietao committed:
level-6: delete (users can now delete tasks from the list), also updated some error handling (mark, unmark and delete, task number provided has to be valid)

joeietao committed:
Level-7: added LoadFile, ReadFile and WriteFile classes

joeietao committed:
update .gitignore

joeietao committed:
update .gitignore

joeietao committed:
update .gitignore

joeietao committed:
Level-8: added Date class

joeietao committed:
Updated .gitignore and added data/duke.txt

joeietao committed:
Merge branch 'branch-Level-7'

joeietao committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

joeietao committed:
fixed bugs relating to the combination of LoadFile and Date

joeietao committed:
made the code more OOP

joeietao committed:
moved all files into the duke package

joeietao committed:
A-JUnit : Add JUnit tests (TodoTest.java and TaskListTest.java)

joeietao committed:
update JUnit Tests

Jacky142857 commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,18 @@
+public class Date {
+    static String formatedDate;
+
+    public static String toString(LocalDate str) {
```
Maybe you can use DateTimeFormatter.ofPattern(pattern) method to make this code shorter

Jacky142857 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,30 @@
+
+            Parser.start();
+
+        } catch (DukeException e) {
```
Maybe you can use 'catch (Exception1 | Exception2 ex)' to catch two exceptions at once

brian16600 replied:
Can catch multiple exceptions if the System.err.print(e) statement is reused

Jacky142857 commented on src/main/java/duke/LoadFile.java:
```diff
@@ -0,0 +1,57 @@
+import java.io.File;
+
+public class LoadFile {
+    public static boolean loaded = false;
```
boolean variable should be name in a way that sound like boolean. For example 'boolean hasLoaded'

Jacky142857 commented on src/main/java/duke/LoadFile.java:
```diff
@@ -0,0 +1,57 @@
+
+                    if (str.contains("[T]")) {
+                        task = new Todo(str.substring(7));
+                    }
```
else should follow the bracket of the previous if block.

Jacky142857 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,108 @@
+    }
+    
+        //to print the current list of items
+        private static void printList(ArrayList<Task> arraylist) {
```
Maybe you should put this method inside Ui class as it deals with user interaction.

Jacky142857 commented on src/main/java/duke/ReadFile.java:
```diff
@@ -0,0 +1,25 @@
+        int n = list.size() - 1;
+        String s = "";
+
+        while (n >= 0) {
```
Maybe you can user a for loop to iterate through the list, instead of using a variable named 'n'.

Dineshraj555 replied:
I agree with @Jacky142857 here

Jacky142857 reviewed (commented):
Overall looks good and you did a good job on text-ui-test. However there are some minor issues with your naming of variables and indentations.

Dineshraj555 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,108 @@
+        String input = scanned.nextLine();
+
+        while (!input.equals(Commands.BYE.command)) {
+            System.out.println("=======================================================================");
```
May this line divider can be declared as a constant static field in UI class and used here instead ?

Dineshraj555 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,38 @@
+        this.date = date;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "[X]" : "[ ]"); // mark done task with X
```
Maybe instead of commenting on the side, you can add javadocs for this method ?

Dineshraj555 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+        System.out.println("Now you have " + Duke.list.size() + " in the list.");
+    }
+
+    protected static void list() {
+
```
whats the reason for this method ?

Dineshraj555 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+        ArrayList<String> arr = new ArrayList<>(Arrays.asList(s.split(" ")));
+        if (arr.contains("/at")) {
+            int i = arr.indexOf("/at");
+            String str1 = String.join(" ", arr.subList(1, i));
+            String str2 = String.join(" ", arr.subList(i + 1, arr.size()));
+            String[] str = new String[2];
+
+
+
+            str[0] = str1;
+            str[1] = str2;
```
maybe you can write a separate private method to pass in the parameters to get your result ? As i notice you had applied the same code for other options of this method. It can make the overall code less repetitive.

Dineshraj555 reviewed (commented):
LGTM. Overall code was easy to understand. I feel the general methods can introduced to make the overall code shorter.

joeietao committed:
A-Jar: create JAR file

brian16600 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+package duke;
+
+import java.time.LocalDate;
+
```
Maybe it might be good to add headers for the constructor class as per the java coding standard

brian16600 commented on src/main/java/duke/LoadFile.java:
```diff
@@ -0,0 +1,57 @@
+
+                    else if (str.contains("[E]")) {
+                        int startIndex = str.indexOf(" (at: ");
+
```
Spacing is a little bit not uniform, perhaps delete some of the empty lines?

brian16600 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,108 @@
+            if (input.equals(Commands.LIST.command)) {
+                printList(Duke.list);
+            }
+
```
It might be better to follow the standard if-else format. Perhaps something like:
} else if (input.length() > 4 && Commands.MARK.command.equals(input.substring(0, 4))) {
                //marks a task as done if the user inputs a string in format "mark xx"
               //NOTE: currently, inputting "mark" alone results in a task named "mark" to be added to the list of tasks
                TaskList.mark(input);
}

joeietao replied:
thank you for the catch! I will update it soon! :)

brian16600 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,108 @@
+
+            //to continue reading user input
+            scanned = new Scanner(System.in);
+            System.out.println("");
```
Consider using \n for the new line instead of an empty line

brian16600 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,42 @@
+//Task class to save description of tasks and mark them as done or undone
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like the naming of the variables here - intuitive and obvious

brian16600 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,42 @@
+    protected String description;
+    protected boolean isDone;
+    protected LocalDate date;
+
```
Similar to previously mentioned, perhaps it may be best to include javadocs comments on the methods in the classes

brian16600 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+            Task toBeMarked = Duke.list.get(Integer.parseInt(input.substring(5)) - 1);
+            toBeMarked.markAsDone();
+        } else {
+            throw new DukeException(
```
There may be no need to split the bracket here since the line itself isnt so long

brian16600 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,199 @@
+        }
+
+        if (stripDescription(input)[1] == "") {
+            throw new DukeException(
```
Perhaps it may be better to do:
throw new DukeException
("Oops, please tell me when is this due!");
As per the java standards

brian16600 reviewed (commented):
Overall the code was quite clean and easy to read, however some code could better follow the java standards/naming conventions. Still very well done overall!

joeietao committed:
Add JavaDocs to the methods and classes

joeietao committed:
fix coding standard

joeietao committed:
add find command, edit TaskList.java and Parser.java

joeietao committed:
Update README.md

Use GFMD in the PR description

joeietao committed:
Update README.md

joeietao committed:
update javadocs

joeietao committed:
Merge branch 'master' of https://github.com/joeietao/ip

# Conflicts:
#	README.md

joeietao committed:
add level 10 and set up gradle

joeietao committed:
update gui

joeietao committed:
built jar file

joeietao committed:
add assertions

there were assumptions in the code so I added assertions

* add assertion to mark command in TaskList
* add assertion to unmark command in TaskList
* add assertion to deadline command in TaskList
* add assertion to todo command in TaskList
* add assertion to event command in TaskList
* add assertion to delete command in TaskList
* add assertion to find command in TaskList

also added exception handling to when user inputs invalid date format for deadline and event commands

joeietao committed:
Merge pull request #2 from joeietao/branch-A-Assertions

add assertions

joeietao committed:
fix code quality

code quality did not follow coding standards given exactly

needed to edit the code to follow the standards

changed:
* naming for variables/methods that could have been named better
* capitalised all constant variables
* edited javadocs from all the changes made
* ensured that there were no magic literals returned from methods

joeietao committed:
Revert "fix code quality"

This reverts commit b412c80424198ce746e92ae603b47aef336f83e3.

joeietao committed:
Merge pull request #3 from joeietao/branch-A-CodeQuality

fix code quality

joeietao committed:
improve on GUI

joeietao committed:
Add Ui.png to docs

joeietao committed:
Add MassOp functionality

Users could only mark, unmark and delete 1 thing at a time, inconvenient and not efficient

The find command could only be used to search for 1 keyword at a time too which may not be very efficient

* added massOp functionality to delete, mark and unmark can now mass delete, mark and unmark.
*also added massOp functionality to find, can now search for multiple keywords at a time

joeietao committed:
update README documentation

joeietao committed:
Update README.md
